### PR TITLE
persist: sharded reads (HollowBatch-based snapshot/listen distribution)

### PR DESCRIPTION
**Pull request by sploiselle**
To enable parallelism in reading from persist, we determined a strategy that involves exchanging `HollowBatch`es across timely workers, and having them download blobs from S3. For more details, see #13884

### Motivation

This PR adds a known-desirable feature. This PR adds the read side of #13884

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - there are no user-facing behavior changes

### CODE REVIEW

File: src/persist-client/src/batch.rs
Comment (danhhz) [inline src/persist-client/src/batch.rs:112]: is this used?

File: src/persist-client/src/read.rs
Comment (danhhz) [inline src/persist-client/src/read.rs:78]: the filtering that we need to do is dependent on what type of batch this is, so I think we should be modeling that explicitly instead of relying on the Some-or-None-ness of frontier

concretely: pull the as_of+frontier into something like an `EnrichedHollowBatchType` enum with one variant for snapshot and one for listen (the corresponding proto should be some sort of proto union). there would be an as_of field on both and only listen would have frontier

File: src/persist-client/src/read.rs
Comment (danhhz) [inline src/persist-client/src/read.rs:128]: i've been imagining that there would be a `fetch_batch` sort of method on ReadHandle that takes in an enriched hollow batch and directly hands back updates and frontier information. but maybe i'm missing something! what does the additional `HollowBatchFetcher` and `fetch` call get us?

File: src/persist-client/src/read.rs
Comment (danhhz) [inline src/persist-client/src/read.rs:430]: let's replace the old codepaths, the new way is strictly better! concretely: replace `Listen::next`'s implementation with this (but keep the name `next`) plz

File: src/persist-client/src/read.rs
Comment (danhhz) [inline src/persist-client/src/read.rs:440]: yeah, this is probably a hint to us that we should find a new name for `frontier` (it was never a great name for what it was doing)

File: src/persist-client/src/read.rs
Comment (danhhz) [inline src/persist-client/src/read.rs:665]: ditto let's replace the old codepaths, the new way is strictly better! concretely: replace `ReadHandle::snapshot`'s implementation with this (but keep the name `snapshot`) plz. then delete SnapshotShard and SnapshotIter entirely

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [inline src/storage/src/source/persist_source.rs:198]: ditto let's replace the old codepaths, the new way is strictly better! concretely: replace `persist_source`'s implementation with this (but keep the name `persist_source`) plz

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [inline src/storage/src/source/persist_source.rs:249]: I think this is copied over from an old impl of persist_source (this call got removed from persist_source at some point last week)

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [inline src/storage/src/source/persist_source.rs:216]: nit: oh because we have two operators now? I'd probably name these according to the operator that moves them or something else a bit more descriptive

it might also "just work" to shadow them with a clone with the same name, move it, and then move the original one into the second operator? (i.e. just put a `let persist_clients = persist_clients.clone();` here)

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [inline src/storage/src/source/persist_source.rs:308]: round robin is what I was thinking, too! :clap:

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [inline src/storage/src/source/persist_source.rs:293]: this is no longer necessary since it's not emitting updates but just low volume batches. we'll definitely need something like this in the second operator though, so maybe this comment is just "move this to the other operator plz"

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [inline src/storage/src/source/persist_source.rs:304]: I'd find this more straightforward to reason about if the stream item was something like `(EnrichedHollowBatch, Option<Antichain<T>>)` instead. maybe? curious what @aljoscha thinks here

it's moving a bit of logic out of persist and into the operators here, but the progress is a function of the stream of batches, not of an individual batch. said another way, if you swapped the order you emitted the batches above so that e.g. you interleaved snapshot and listen batches, the way this code is now would be incorrect and that wouldn't be super obvious. at least if we were doing the progress stuff above, it would be obvious that there was _some_ contract being upheld up there if someone went in and messed with the batch order

perhaps even better is to make a new `Subscription` struct in persist that looks a lot like Listen and is a thin wrapper for both a snapshot _and_ a Listen. then we can move all the frontier logic there

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [inline src/storage/src/source/persist_source.rs:355]: note to self: we'll need to think carefully through the since capabilities of both of these operators, to make sure everything is downgrading correctly

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [inline src/storage/src/source/persist_source.rs:378]: not 100% sure I understand the question but if it helps, the two operators have different capabilities. you can think of the capabilities of the first stage as something like "here are all the batches one needs to fetch to progress to this point" and the second one like "here are all the actual fetched updates needed to progress to this point"

in fact, I think this suggests we have an opportunity to tighten up the contract between the two stages here:
- the first operator should emit batches and use its output capability to communicate when the full set of batches for some time have been emitted (as it does now)
- this one simply needs to flat_map the batches out by fetching them. concretely: if this operator sees "B1, B2, F1, B3, F2", it simply needs to fetch and emit B1 and B2 before passing on F1 (via cap.downgrade), similarly with B3 and F2. if we decide to in the future, it's safe to begin fetching B3 before we've finished fetching and emitting B1 and B2, we just have to make sure the progress info respects the original order that the batches and frontiers came in
- concretely: this means `HollowBatchFetcher::fetch` (or `ReadHandle::fetch_batch` if we do my comment above) just needs to return updates. it doesn't need to have anything to do with progress or frontiers

**Review by danhhz — COMMENTED** [review]
the skeleton/approach looks great! the most substantive of my comments was a series of "let's replace the old impl with the new one plz"

I also suspect we also have a bit of a persist-wide FooBatch naming bikeshed coming our way (i'm also thinking ahead to the write split stuff). let's get you, me, @aljoscha, and @pH14 on a call to hash it out and, in the meantime, leave you Batch naming here as-is

File: src/persist-client/src/batch.rs
Comment (sploiselle) [reply src/persist-client/src/batch.rs:112]: This is just junk from debugging.

File: src/persist-client/src/read.rs
Comment (sploiselle) [reply src/persist-client/src/read.rs:78]: Yep-this is a TODO elsewhere (probably should have been a TODO here)

File: src/persist-client/src/read.rs
Comment (sploiselle) [reply src/persist-client/src/read.rs:128]: Can get reduced away, I'm almost certain.

File: src/persist-client/src/read.rs
Comment (sploiselle) [reply src/persist-client/src/read.rs:430]: 👌

File: src/persist-client/src/read.rs
Comment (sploiselle) [reply src/persist-client/src/read.rs:440]: Nice--had the same thought here.

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs:249]: Yeah, this hasn't been rebased yet bc I wanted to make sure the approach was sane before taking on that endeavor.

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs:304]: Oh, that is great and so sensible. Will do.

File: src/storage/src/source/persist_source.rs
Comment (aljoscha) [reply src/storage/src/source/persist_source.rs:378]: yes, I think frontier/progress tracking is only a concern of the first-stage operator

File: src/storage/src/source/persist_source.rs
Comment (frankmcsherry) [reply src/storage/src/source/persist_source.rs:308]: Minorest of nits: round robin occasionally gets itself in trouble with self-synchonization: you have a few sources running, and they all use the same strategy, and .. surprisingly often they end up working in lock step. So, if you have 10 sources, each time slice one worker gets 10 batches rather than distributing them (the round robin counter isn't shared across sources). Easy-ish fix would be to use `(source_id, i).hashed() % peers` or something like it!

File: src/storage/src/source/persist_source.rs
Comment (frankmcsherry) [reply src/storage/src/source/persist_source.rs:355]: If it helps, I think the second operator is morally a `flat_map`, albeit one backed by S3 rather than a function. In that case, as long as a hollow batch is transmitted with a timestamp corresponding to its `lower` frontier, the mostly-vanilla timely idioms should work (hold the capability you receive the hollow batch with, until it has been completely produced).

I'm now reading the code though, and inclined to propose that, unless there are strong negative feels (in which case, no worries). If the *first* operator receives batch descriptions and progress statements, it can forward the batch descriptions to workers, and hold back its capability based on the progress statements. The *second* operators then don't need to do anything about capabilities (frontiers are held back by the first operator holding the capability).

Lmk if this is confusing or wrong. I think it makes a bunch of sense, but also I'm not sure if there are `persist` constraints that make sending a stream of `HollowBatch` things easier than `Enum(HollowBatch, Progress)`.

File: src/storage/src/source/persist_source.rs
Comment (frankmcsherry) [reply src/storage/src/source/persist_source.rs:378]: Ah, this seems to be the comments I posted without reading all the other comments first. I agree with these takes, and my take is purely redundant with these.

Comment (frankmcsherry) [reply src/storage/src/source/persist_source.rs:378]: > concretely: if this operator sees "B1, B2, F1, B3, F2", it simply needs to fetch and emit B1 and B2 before passing on F1 (via cap.downgrade), 

I'm not sure the second operator needs to do `cap.downgrade` at all! It probably just grabs the capability that came with e.g. `B1`, `B2`, and uses it to send the updates for each of the batches, and then just drops the capability. It only needs to hold and downgrade the capability if it plans to emit output updates at times that it will not receive in its input (and I think that is not the case here?).

Comment (aljoscha) [reply src/storage/src/source/persist_source.rs:378]: > my take is purely redundant with these.

Still good to see that everyone arrived at the same conclusion! :sweat_smile:

**Comment by sploiselle** [comment]
I'm just doing a CI run of my local changes; this is still in draft and requires more responses to everyone's feedback.

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [reply src/storage/src/source/persist_source.rs]: > I'm not sure the second operator needs to do cap.downgrade at all! It probably just grabs the capability that came with e.g. B1, B2, and uses it to send the updates for each of the batches, and then just drops the capability. It only needs to hold and downgrade the capability if it plans to emit output updates at times that it will not receive in its input (and I think that is not the case here?).

You are correct of course! I'm so used to only writing source operators at this point that I brain-farted and forgot that incoming data has an attached capability.

File: src/persist-client/src/read.rs
Comment (sploiselle) [reply src/persist-client/src/read.rs]: To keep this PR as small as possible, I'll do this in a follow up. I have the code locally, just don't want to gum this one up any more than necessary.

File: src/persist-client/src/read.rs
Comment (sploiselle) [reply src/persist-client/src/read.rs]: I came up with the name `until` instead of `frontier` in the context that this is used (as the reader metadata). idk if that's true/useful/idiomatic, but made some sense to me. very open to suggestions.

File: src/persist-client/src/read.rs
Comment (sploiselle) [reply src/persist-client/src/read.rs]: ditto on follow-up PR

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs]: ditto on follow-up PR

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs]: Came up with clearer names

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs:126]: Implemented Frank's suggestions here.

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs:122]: I put this in a new `Subscribe` struct; lmk if that was what you were thinking.

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [inline src/storage/src/source/persist_source.rs:142]: will fix this on next push

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [inline src/storage/src/source/persist_source.rs:150]: will fix this name on next push

**Comment by sploiselle** [comment]
@danhhz @aljoscha ptal--there are some minor bits I missed and will clean up on my next pass through CI.

File: src/persist-client/src/read.rs
Comment (danhhz) [reply src/persist-client/src/read.rs]: Hmmm can you at least push that local branch somewhere I can look at it then? As-is, I have to manually eyeball the old impl vs the new one to review it (instead of git doing that for me 😄).

Concretely, stuff like the `read.downgrade_since(as_of_moveable.clone()).await;` comment I left in the last round would jump out if the old impl was replaced by the new one

Comment (sploiselle) [reply src/persist-client/src/read.rs]: Pushed all of that code up; makes the diff much noisier the way I have it. lmk if it's too chaotic and I can reorganize the commits to make it more legible.

**Comment by sploiselle** [comment]
Forgot to rebase on `main` to fix the failing PG CDC test, but that's a known issue I can fix on next push.

File: src/storage/src/source/persist_source.rs
Comment (aljoscha) [inline src/storage/src/source/persist_source.rs:42]: :raised_hands:

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [reply src/storage/src/source/persist_source.rs:127]: Yeah, this is great!

File: src/persist-client/src/read.rs
Comment (danhhz) [inline src/persist-client/src/read.rs:548]: I think Subscribe::new probably wants to take the Vec of batches + Listen, so we don't have to duplicate stuff like this `verify_listen` call (so just move the `read.snapshot` and `read.listen` call that you currently have in `new` to here)

**Comment by danhhz** [comment]
oh, also, can we update the PR title and description, please? :)

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs:180]: That is what [this code](https://github.com/MaterializeInc/materialize/pull/13849/files#diff-6f304b700fb56de26b66e09ff27e8c3dcea8bfff6aea64da3c291a4a6b24b369R616-R618) in `fetch_batch` attempts to do. Something I missed here?

Comment (danhhz) [reply src/storage/src/source/persist_source.rs:180]: ah! that feels to me like a confusing contract for fetch_batch. (partial evidence: it's surprising enough to me that I didn't even realize it was happening while looking over this PR)

Comment (sploiselle) [reply src/storage/src/source/persist_source.rs:180]: Is it just the name of the function that you drew assumptions from? Would something like `process_batch` be more illustrative from your perspective?

Comment (danhhz) [reply src/storage/src/source/persist_source.rs:180]: no, I think it's just an odd contract to have for a method that exists on ReadHandle, which is pretty general purpose. 
I let this mull for the last hour or so and I think everything clears up if we have a new something (BatchStreamFetcher?) that wraps the ReadHandle and this `fetch_batch` method is only available on that (think the pattern that Listen has now)

I actually think I'd like to introduce something exactly along those lines so that we can abstract over pipelining batch fetches, etc. that should 100% be a followup to this PR, so, in conclusion, I think what you have now is fine to merge as-is. do me a favor a leave a TODO next to that code you linked above about it potentially being surprising for a ReadHandle user, please

@aljoscha one thing I realized mulling all this over is that our story around the seqno capabilities is a little underdeveloped here. the first operator is handing our references to things, how do we ensure the second operator doesn't have those things gc'd out from under it?

**Comment by sploiselle** [comment]
Just noting that I have all of @danhhz's comments implemented locally and passing tests, but am waiting for @aljoscha's feedback on https://github.com/MaterializeInc/materialize/pull/13849#discussion_r934856526

File: src/storage/src/source/persist_source.rs
Comment (aljoscha) [reply src/storage/src/source/persist_source.rs:180]: I think we might not currently be ensuring that! The immediate solution that comes to mind is that persist does ref-counting for capabilities, similar to how timely does. The first stage operator (the one that listens for changes) would records that it hands out a capability for a given seqno, the reader operator then records that it is done with it, reducing the ref count. 

This is just a strawman, for now!

Comment (aljoscha) [reply src/storage/src/source/persist_source.rs:180]: More spitballing: The "subscribe" operator is the only one that really cares about holding back the `since` frontier. The "fetcher" operators don't care about the `since` frontier at all but they _do_ care about holding back the `SeqNo` frontier.

What we could do is: the fetcher operators turn their `ReadHandle` into a special `FetchHandle`, which immediately let's the `since` go to `[]`. We track the frontier of the required `SeqNo` from the subscribe operator to the fetch operator and the fetcher operators _explicitly_ downgrade the `SeqNo` frontier using the `FetchHandle`.

In state, the `FetchHandle` could look like a regular `ReadHandle` that just has it's `since` at `[]` or we could track them separately in state. I'm sure we all don't want the added complexity of the latter, though.

File: src/persist-client/src/read.rs
Comment (sploiselle) [reply src/persist-client/src/read.rs]: TIL

File: src/storage/src/controller.rs
Comment (sploiselle) [reply src/storage/src/controller.rs]: Added `snapshot_and_fetch`

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs]: This must be a `u64`, which I had just mindlessly typed as `Timestamp`; "fixed" this

**Comment by sploiselle** [comment]
@danhhz ready for review again--I've seen mention that the failing tests are flaky, so confident that they aren't failing due to the submitted diff

File: src/storage/src/source/persist_source.rs
Comment (danhhz) [reply src/storage/src/source/persist_source.rs:178]: How would the ref count work? We can't use an Arc because it has to be ExchangeData. Are you thinking of tracking it state? That would be pretty expensive and unfortunate.

I like something like the FetchHandle, I've been imagining something vaguely like it to abstract over pipelining the batch fetches, anyway. 100% not suggesting adding FetchHandle as part of a scope increase for sean in this PR. But, given that things are working right now in main, I'm a little hesitant to merge and potentially create another fire for persist without 1. being comfortable with the design for how we'll fix this and 2. having someone lined up to do it as a fast-follow. I also don't want to leave this PR open to rot if we don't have someone right away for (2).

@aljoscha can we sync up on what we want to do here as soon as I'm on tomorrow?

**Review by danhhz — APPROVED** [review]
LGTM modulo the outstanding seqno capability discussion. Great work!

File: src/storage/src/source/persist_source.rs
Comment (sploiselle) [reply src/storage/src/source/persist_source.rs:178]: fwiw this is my top priority and if there is follow-on work to make this Good™, I am glad to do it. I don't think adding a `FetchHandle` is much work and I'm glad to make it part of this PR, or an immediate follow up.

Comment (danhhz) [reply src/storage/src/source/persist_source.rs:178]: :thumbsup: I think I do need to have the design discussion with aljoscha before we can make any concrete decisions on next steps

**Comment by sploiselle** [comment]
@danhhz last commit adds signatures we chatted about

**Review by danhhz — APPROVED** [review]
merge away!

**Comment by aljoscha** [comment]
🙌

**Pull request merged**
